File: src/depth5_sweep.cpp
#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx2,fma")
// Depth-5 sweep: threshold landscape of smallest Leontovich graphs
// among all spherically symmetric trees T(d1,...,dk) for k ≤ 5.
//
// Uses EXACT integer arithmetic (__int128) on the (k+1)×(k+1)
// similarity matrix.  Falls back to extended range via log-space
// sign detection when __int128 overflows.
//
// Compile: g++ -O3 -march=native -std=c++17 -fopenmp -o depth5_sweep
//          src/depth5_sweep.cpp
// Usage:   ./depth5_sweep [--max-vertices 400] [--max-degree 20]
//                         [--max-depth 5]

#include <algorithm>
#include <atomic>
#include <cstdio>
#include <cstring>
#include <map>
#include <string>
#include <vector>

static FILE* log_fp = nullptr;

static constexpr int MAX_DIM = 7;  // max orbits = max_depth + 1
static constexpr int MAX_N = 301;  // max path length to check

using i128 = __int128;

// Overflow-checked multiply using compiler intrinsic.
// __builtin_mul_overflow is immune to UB-based optimization removal.
static inline bool safe_mul(i128 a, i128 b, i128& result) {
    return !__builtin_mul_overflow(a, b, &result);
}

// Overflow-checked add using compiler intrinsic.
static inline bool safe_add(i128 a, i128 b, i128& result) {
    return !__builtin_add_overflow(a, b, &result);
}

struct SimilarityMatrix {
    int dim;
    int M[MAX_DIM][MAX_DIM];  // integer quotient matrix entries
    int a[MAX_DIM];           // orbit sizes
    int total_v;

    void build(const std::vector<int>& degrees) {
        dim = (int)degrees.size() + 1;
        memset(M, 0, sizeof(M));
        for (int i = 0; i < (int)degrees.size(); i++) {
            M[i][i + 1] = degrees[i];
            M[i + 1][i] = 1;
        }
        a[0] = 1;
        for (int i = 0; i < (int)degrees.size(); i++)
            a[i + 1] = a[i] * degrees[i];
        total_v = 0;
        for (int i = 0; i < dim; i++) total_v += a[i];
    }
};

struct LeoResult {
    bool found;
    int first_n;
    int first_d;
    double ratio;
};

static LeoResult check_leontovich(const SimilarityMatrix& S) {
    int dim = S.dim;

    // v[step][orbit] = (M^step · 1)[orbit] -- exact integers
    static thread_local i128 v[MAX_N][MAX_DIM];
    static thread_local i128 homP[MAX_N + 1];

    memset(v, 0, sizeof(i128) * MAX_N * MAX_DIM);
    memset(homP, 0, sizeof(homP));

    for (int i = 0; i < dim; i++) v[0][i] = 1;
    homP[1] = S.total_v;

    int safe_steps = MAX_N;  // last step where no overflow occurred

    for (int step = 1; step < MAX_N; step++) {
        bool overflow = false;
        for (int i = 0; i < dim; i++) {
            i128 s = 0;
            for (int j = 0; j < dim; j++) {
                i128 prod;
                if (!safe_mul((i128)S.M[i][j], v[step - 1][j], prod)) {
                    overflow = true;
                    break;
                }
                if (!safe_add(s, prod, s)) {
                    overflow = true;
                    break;
                }
            }
            if (overflow) break;
            v[step][i] = s;
        }
        if (overflow) {
            safe_steps = step;
            break;
        }

        i128 h = 0;
        for (int i = 0; i < dim; i++) {
            i128 term;
            if (!safe_mul((i128)S.a[i], v[step][i], term) ||
                !safe_add(h, term, h)) {
                safe_steps = step;
                goto done_powers;
            }
        }
        homP[step + 1] = h;
    }
done_powers:

    // Check crossovers using exact arithmetic
    for (int d = 2; d < 21; d++) {
        for (int stem = 0; stem < safe_steps - d - 1; stem++) {
            int n = stem + d + 2;
            if (n >= safe_steps + 1) break;  // need homP[n] which uses step n-1
            if (homP[n] <= 0) continue;

            // homE = sum_i a[i] * v[stem][i] * v[1][i] * v[d][i]
            i128 homE = 0;
            bool overflow = false;
            for (int i = 0; i < dim; i++) {
                i128 t1, t2, t3;
                if (!safe_mul(v[stem][i], v[1][i], t1) ||
                    !safe_mul(t1, v[d][i], t2) ||
                    !safe_mul((i128)S.a[i], t2, t3) ||
                    !safe_add(homE, t3, homE)) {
                    overflow = true;
                    break;
                }
            }
            if (overflow) break;  // can't verify at this d,n

            // Exact comparison: is homE < homP[n]?
            if (homE < homP[n]) {
                // Genuine Leontovich crossover (exact)
                double ratio =
                    (double)homE / (double)homP[n];  // approx for display
                return {true, n, d, ratio};
            }
        }
    }
    return {false, 0, 0, 0};
}

struct FrontierEntry {
    int total_v;
    std::vector<int> degrees;
    int d;
    double ratio;
};

static int max_vertices = 400;
static int max_degree = 20;
static int max_depth = 5;
static std::atomic<long long> total_checked{0};
static std::atomic<long long> total_leo{0};

// Key: (threshold_n, orbit_count)
using FKey = std::pair<int, int>;
static std::map<FKey, FrontierEntry> frontier;

// Thread-local recursive sweep (no global mutation except atomics)
static void sweep(std::vector<int>& current, int remaining, int orbit_product,
                  int current_v, long long& local_checked, long long& local_leo,
                  std::map<FKey, FrontierEntry>& local_frontier) {
    if (remaining == 0) {
        local_checked++;

        if (local_checked % 100000 == 0) {
            fprintf(stderr, "\r    ... checked %lld (local), found %lld Leo   ",
                    local_checked, local_leo);
        }

        SimilarityMatrix S;
        S.build(current);

        auto result = check_leontovich(S);
        if (result.found) {
            local_leo++;
            int n = result.first_n;
            int orbits = (int)current.size() + 1;
            FKey key = {n, orbits};
            if (local_frontier.find(key) == local_frontier.end() ||
                S.total_v < local_frontier[key].total_v) {
                local_frontier[key] = {S.total_v, current, result.first_d,
                                       result.ratio};
            }
#pragma omp critical
            {
                if (frontier.find(key) == frontier.end() ||
                    S.total_v < frontier[key].total_v) {
                    frontier[key] = {S.total_v, current, result.first_d,
                                     result.ratio};
                    printf("  NEW BEST n>=%d [%d-orb]: T(", n, orbits);
                    for (int i = 0; i < (int)current.size(); i++) {
                        if (i) printf(",");
                        printf("%d", current[i]);
                    }
                    printf(") |V|=%d d=%d r=%.8f\n", S.total_v, result.first_d,
                           result.ratio);
                    fflush(stdout);
                    if (log_fp) {
                        fprintf(log_fp, "n>=%d,%d-orb,T(", n, orbits);
                        for (int i = 0; i < (int)current.size(); i++) {
                            if (i) fprintf(log_fp, ",");
                            fprintf(log_fp, "%d", current[i]);
                        }
                        fprintf(log_fp, "),%d,%d,%.8f\n", S.total_v,
                                result.first_d, result.ratio);
                        fflush(log_fp);
                    }
                }
            }
        }
        return;
    }

    for (int d = 1; d <= max_degree; d++) {
        int new_v = current_v + orbit_product * d;
        if (new_v > max_vertices) break;
        current.push_back(d);
        sweep(current, remaining - 1, orbit_product * d, new_v, local_checked,
              local_leo, local_frontier);
        current.pop_back();
    }
}

// Merge a local frontier into the global one (called under critical)
static void merge_frontier(const std::map<FKey, FrontierEntry>& local) {
    for (auto& [key, e] : local) {
        if (frontier.find(key) == frontier.end() ||
            e.total_v < frontier[key].total_v) {
            frontier[key] = e;
        }
    }
}

int main(int argc, char** argv) {
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--max-vertices" && i + 1 < argc)
            max_vertices = std::stoi(argv[++i]);
        else if (arg == "--max-degree" && i + 1 < argc)
            max_degree = std::stoi(argv[++i]);
        else if (arg == "--max-depth" && i + 1 < argc)
            max_depth = std::stoi(argv[++i]);
        else if (arg == "--log" && i + 1 < argc) {
            log_fp = fopen(argv[++i], "w");
            if (!log_fp) {
                fprintf(stderr, "Cannot open log file: %s\n", argv[i]);
                return 1;
            }
            fprintf(log_fp, "threshold,tree,vertices,d,ratio\n");
            fflush(log_fp);
        } else if (arg == "-h" || arg == "--help") {
            printf(
                "Usage: %s [--max-vertices N] [--max-degree N] [--max-depth "
                "N] [--log FILE]\n",
                argv[0]);
            return 0;
        }
    }

    printf("=== Depth-%d Leontovich Sweep (C++, __int128 exact) ===\n",
           max_depth);
    printf("Max vertices: %d, Max degree: %d\n\n", max_vertices, max_degree);

    setvbuf(stderr, NULL, _IONBF, 0);

    for (int depth = 1; depth <= max_depth; depth++) {
        fprintf(stderr, "  [depth %d] sweeping...\n", depth);

        // Parallelize over the first branching degree d1
        int d1_max = std::min(max_degree, max_vertices - 1);

#pragma omp parallel
        {
            long long local_checked = 0, local_leo = 0;
            std::map<FKey, FrontierEntry> local_front;

#pragma omp for schedule(dynamic, 1) nowait
            for (int d1 = 1; d1 <= d1_max; d1++) {
                int v1 = 1 + d1;  // root + d1 children
                if (v1 > max_vertices) continue;
                if (depth == 1) {
                    // Single-level tree: just T(d1)
                    std::vector<int> seq = {d1};
                    local_checked++;
                    SimilarityMatrix S;
                    S.build(seq);
                    auto result = check_leontovich(S);
                    if (result.found) {
                        local_leo++;
                        int n = result.first_n;
                        FKey key = {n, 2};  // depth-1 = 2 orbits
                        if (local_front.find(key) == local_front.end() ||
                            S.total_v < local_front[key].total_v) {
                            local_front[key] = {S.total_v, seq, result.first_d,
                                                result.ratio};
                        }
                    }
                } else {
                    // Recurse for remaining (depth-1) levels
                    std::vector<int> seq = {d1};
                    sweep(seq, depth - 1, d1, v1, local_checked, local_leo,
                          local_front);
                }

                // Progress (only from first thread hitting this d1)
                int pct =
                    (int)((long long)(d1 - 1) * 100 / std::max(1, d1_max));
                fprintf(
                    stderr,
                    "\r  [depth %d] d1=%d/%d (%d%%)  checked=%lld  leo=%lld",
                    depth, d1, d1_max, pct,
                    total_checked.load(std::memory_order_relaxed) +
                        local_checked,
                    total_leo.load(std::memory_order_relaxed) + local_leo);
            }

#pragma omp critical
            {
                total_checked += local_checked;
                total_leo += local_leo;
                merge_frontier(local_front);
            }
        }

        fprintf(stderr,
                "\r  [depth %d] done.                                  \n",
                depth);
    }

    printf("\n============================================================\n");
    printf("Checked %lld branching sequences, found %lld Leontovich\n",
           total_checked.load(), total_leo.load());
    printf("\n============================================================\n");
    printf(
        "THRESHOLD LANDSCAPE: Smallest Leontovich graph per n (all orbits)\n");
    printf("============================================================\n");

    // Group by n
    std::map<int, std::map<int, FrontierEntry>> by_n;
    for (auto& [key, e] : frontier) {
        auto [n, orbits] = key;
        by_n[n][orbits] = e;
    }

    int abs_min_n = -1, abs_min_v = 1 << 30;
    for (auto& [n, orb_map] : by_n) {
        printf("n>=%3d:", n);
        for (auto& [orbits, e] : orb_map) {
            printf("  T(");
            for (int i = 0; i < (int)e.degrees.size(); i++) {
                if (i) printf(",");
                printf("%d", e.degrees[i]);
            }
            printf("),v=%d", e.total_v);
            if (e.total_v < abs_min_v) {
                abs_min_v = e.total_v;
                abs_min_n = n;
            }
        }
        printf("\n");
    }

    if (abs_min_n >= 0) {
        printf("\nAbsolute minimum |V|=%d at n>=%d\n", abs_min_v, abs_min_n);
    }

    if (log_fp) fclose(log_fp);
    return 0;
}

File: scripts/depth5_sweep.py
#!/usr/bin/env python3
"""
Depth-5 sweep: find the threshold landscape of smallest Leontovich graphs
among all spherically symmetric trees T(d1, d2, ..., dk) for k ≤ 5.

Uses the O(1) similarity matrix trick: for a tree with branching
sequence D = [d1, ..., dk], the quotient matrix M is (k+1)×(k+1)
tridiagonal. We compute eigenvalues of M and use the orbit-weighted
transfer matrix to evaluate hom(P_n, H) and hom(E_n, H) in O(k²)
per n, avoiding the full |V|×|V| adjacency matrix entirely.

Usage:
    python3 scripts/depth5_sweep.py [--max-vertices 400] [--max-degree 20]
"""

import argparse
import sys

import numpy as np


def get_similarity_matrix(degrees):
    """
    Build similarity matrix M and orbit sizes for symmetric tree T(d1, ..., dk).

    M is (k+1)×(k+1) tridiagonal:
      M[i, i+1] = d_{i+1}  (parent → children)
      M[i+1, i] = 1        (child → parent)

    Orbit sizes: a = [1, d1, d1*d2, d1*d2*d3, ...]
    Total vertices: sum(a)
    """
    k = len(degrees)
    dim = k + 1
    M = np.zeros((dim, dim))
    for i in range(k):
        M[i, i + 1] = degrees[i]
        M[i + 1, i] = 1

    a = [1]
    for d in degrees:
        a.append(a[-1] * d)

    return M, a


def check_leontovich_similarity(M, a, max_n=300):
    """
    Check Leontovich property using the similarity matrix.

    We compute w[step] = M^step · 1_weighted, where the initial vector
    is the orbit sizes (since each orbit contributes a[i] identical vertices).

    Returns (is_leo, first_n, first_d, ratio, lam1, lam2) or None.
    """
    dim = len(a)
    total = sum(a)

    # Weighted walk counts: w[step][i] = (A^step · 1)[any vertex in orbit i]
    # homP[n] = sum_i a[i] * w[n-1][i]
    # For the similarity matrix, we iterate w = M @ w but need to account
    # for the orbit weighting properly.
    #
    # The trick: let v[0] = vector of 1s (one per orbit).
    # Then v[step] = M^step @ v[0].
    # hom(P_n, H) = sum_i a[i] * v[n-1][i]
    #
    # For E_n with depth d: hom(E_n, H) = sum_i a[i] * v[stem][i] * b[i]
    # where b[i] = v[1][i] * v[d][i] (the branch contribution)

    v = np.zeros((max_n, dim))
    v[0] = np.ones(dim)
    a_arr = np.array(a, dtype=np.float64)

    homP = np.zeros(max_n + 1)
    homP[1] = total

    for step in range(1, max_n):
        v[step] = M @ v[step - 1]
        homP[step + 1] = np.dot(a_arr, v[step])

    # Eigenvalues for reporting
    evals = sorted(np.linalg.eigvalsh(M), reverse=True)
    lam1 = evals[0] if len(evals) > 0 else 0
    lam2 = evals[1] if len(evals) > 1 else 0

    # Check crossovers for each depth d
    for d in range(2, min(21, max_n - 2)):
        b = v[1] * v[d]
        for stem in range(0, max_n - d - 2):
            homE = np.dot(a_arr, v[stem] * b)
            n = stem + d + 2
            if n >= max_n:
                break
            if homP[n] > 0 and homE / homP[n] < 1.0 - 1e-11:
                ratio = homE / homP[n]
                return True, n, d, ratio, lam1, lam2

    return False, None, None, None, lam1, lam2


def enumerate_branching_sequences(max_depth, max_degree, max_vertices):
    """
    Yield all branching sequences [d1, ..., dk] for k=1..max_depth
    where each di ∈ [1, max_degree] and total vertices ≤ max_vertices.
    """
    for depth in range(1, max_depth + 1):
        # Generate sequences with pruning
        yield from _enumerate_recursive(depth, max_degree, max_vertices, [])


def _enumerate_recursive(remaining, max_degree, max_vertices, current):
    """Recursively enumerate branching sequences with early pruning."""
    if remaining == 0:
        yield tuple(current)
        return

    # Current vertex count
    orbit_product = 1
    for d in current:
        orbit_product *= d
    current_vertices = sum(_prefix_product(current, i) for i in range(len(current) + 1))

    for d in range(1, max_degree + 1):
        # New orbit adds orbit_product * d vertices
        new_vertices = current_vertices + orbit_product * d
        if new_vertices > max_vertices:
            break
        current.append(d)
        yield from _enumerate_recursive(
            remaining - 1, max_degree, max_vertices, current
        )
        current.pop()


def _prefix_product(seq, length):
    """Product of first `length` elements of seq."""
    p = 1
    for i in range(length):
        p *= seq[i]
    return p


def main():
    parser = argparse.ArgumentParser(
        description="Sweep spherically symmetric trees up to depth 5"
    )
    parser.add_argument(
        "--max-vertices",
        type=int,
        default=400,
        help="Maximum vertex count (default: 400)",
    )
    parser.add_argument(
        "--max-degree",
        type=int,
        default=20,
        help="Maximum branching degree (default: 20)",
    )
    parser.add_argument(
        "--max-depth",
        type=int,
        default=5,
        help="Maximum tree depth / branching sequence length (default: 5)",
    )
    args = parser.parse_args()

    print(f"=== Depth-{args.max_depth} Leontovich Sweep ===")
    print(f"Max vertices: {args.max_vertices}, Max degree: {args.max_degree}")
    print()

    # Track: for each threshold n, the smallest |V| Leontovich graph
    frontier = {}  # threshold_n -> (|V|, degrees, d, ratio, lam1, lam2)
    total_checked = 0
    total_leo = 0

    for degrees in enumerate_branching_sequences(
        args.max_depth, args.max_degree, args.max_vertices
    ):
        total_checked += 1

        M, a = get_similarity_matrix(degrees)
        total_v = sum(a)

        is_leo, first_n, first_d, ratio, lam1, lam2 = check_leontovich_similarity(M, a)

        if is_leo:
            total_leo += 1
            if first_n not in frontier or total_v < frontier[first_n][0]:
                frontier[first_n] = (total_v, degrees, first_d, ratio, lam1, lam2)
                print(
                    f"  NEW BEST n≥{first_n}: T{degrees} |V|={total_v} "
                    f"d={first_d} r={ratio:.8f} λ1={lam1:.4f} λ2={lam2:.4f}"
                )

        if total_checked % 10000 == 0:
            print(
                f"  ... checked {total_checked} sequences, "
                f"found {total_leo} Leontovich ...",
                file=sys.stderr,
            )

    print(f"\n{'=' * 60}")
    print(f"Checked {total_checked} branching sequences, found {total_leo} Leontovich")
    print(f"\n{'=' * 60}")
    print("THRESHOLD LANDSCAPE: Smallest Leontovich graph per threshold n")
    print(f"{'=' * 60}")
    print(
        f"{'n':>4s}  {'|V|':>6s}  {'Structure':25s}  {'d':>3s}  {'λ1':>8s}  {'λ2':>8s}  {'ratio':>12s}"
    )
    print("-" * 72)

    for n in sorted(frontier.keys()):
        v, deg, d, ratio, lam1, lam2 = frontier[n]
        deg_str = f"T{deg}"
        print(
            f"{n:4d}  {v:6d}  {deg_str:25s}  {d:3d}  {lam1:8.4f}  {lam2:8.4f}  {ratio:12.8f}"
        )

    # Also find absolute minimum
    if frontier:
        min_n, (min_v, min_deg, min_d, min_r, min_l1, min_l2) = min(
            frontier.items(), key=lambda x: x[1][0]
        )
        print(f"\nAbsolute minimum: T{min_deg} with |V|={min_v} at n≥{min_n}")


if __name__ == "__main__":
    main()
